heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/metricscache/CacheCore.java

```diff
@@ -0,0 +1,543 @@
+    }
+    Map<String, Integer> map = idxComponentInstance.get(componentName);
+    if (!map.containsKey(instanceId)) {
+      map.put(instanceId, componentInstanceCount++);
+    }
+  }
+
+  public boolean existComponentInstance(String componentName, String instanceId) {
+    if (componentName != null
+        && !idxComponentInstance.containsKey(componentName)) {
+      return false;
+    }
+    if (instanceId != null
+        && !idxComponentInstance.get(componentName).containsKey(instanceId)) {
+      return false;
+    }
+    return true;
+  }
+
+  public boolean existMetric(String name) {
+    return idxMetricName.containsKey(name);
+  }
+
+  private void assureMetricName(String name) {
+    if (!idxMetricName.containsKey(name)) {
+      idxMetricName.put(name, metricNameCount++);
+    }
+  }
+
+  /**
+   * compatible with heron::tmaster::TMetricsCollector
+   */
+  public void addMetricException(TopologyMaster.PublishMetrics metrics) {
+    synchronized (CacheCore.class) {
+      for (TopologyMaster.MetricDatum metricDatum : metrics.getMetricsList()) {
+        addMetric(metricDatum);
+      }
+      for (TopologyMaster.TmasterExceptionLog exceptionLog : metrics.getExceptionsList()) {
+        addException(exceptionLog);
+      }
+    }
+  }
+
+  /**
+   * make bucket id
+   *
+   * @param hi index of [component-instance]
+   * @param lo index of metric name
+   * @return bucket id
+   */
+  private long makeBucketId(int hi, int lo) {
+    return (((long) hi) << 32) | (lo & 0xffffffffL);
+  }
+
+  private void addMetric(TopologyMaster.MetricDatum metricDatum) {
+    String componentName = metricDatum.getComponentName();
+    String instanceId = metricDatum.getInstanceId();
+    String metricName = metricDatum.getName();
+    LOG.info(componentName + " " + instanceId + " " + metricName);
+    assureComponentInstance(componentName, instanceId);
+    assureMetricName(metricName);
+    // calc bucket idx
+    int idx1 = idxComponentInstance.get(componentName).get(instanceId);
+    int idx2 = idxMetricName.get(metricName);
+    long bucketId = makeBucketId(idx1, idx2);
+    LOG.info(Long.toHexString(bucketId) + " = "
+        + Integer.toHexString(idx1) + " " + Integer.toHexString(idx2));
+    // fetch the bucket
+    Map.Entry<Long, Map<Long, LinkedList<MetricDatapoint>>> entry =
+        cacheMetric.floorEntry(metricDatum.getTimestamp());
+    if (entry != null) {
+      Map<Long, LinkedList<MetricDatapoint>> locator = entry.getValue();
+      if (!locator.containsKey(bucketId)) {
+        locator.put(bucketId, new LinkedList<MetricDatapoint>());
+      }
+      LinkedList<MetricDatapoint> bucket = locator.get(bucketId);
+      // store the metric
+      MetricDatapoint datum = new MetricDatapoint();
+      datum.timestamp = metricDatum.getTimestamp();
+      datum.value = metricDatum.getValue();
+      bucket.offerFirst(datum);
+      LOG.info("addMetrics " + datum);
+    } else {
+      LOG.warning("too old metric: " + metricDatum);
```

Comment:
added more context.

if the metric falls out of cache timestamp window, it is ignored.